### ISSUE TITLE
Performance (time) regression starting in OpenJ9-0.16.0, not a problem in 0.15.1

### ISSUE THREAD

**Issue by tommyettinger** [eclipse/openj9#7525]
Java -version output
--------------------

I'm on Windows 7 with a 6th-gen i7 quad mobile processor and 16GB DDR4 RAM.

This version shows the regression:

# "C:\Program Files\AdoptOpenJDK\jdk-13.0.0.33-openj9\bin\java.exe" -version
openjdk version "13" 2019-09-17
OpenJDK Runtime Environment AdoptOpenJDK (build 13+33)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.16.0, JRE 13 Windows 7 amd64-64-Bit Compressed References 20190916_74 (JIT enabled, AOT enabled)
OpenJ9   - 867dab457
OMR      - d4c85c31
JCL      - 2858e3f001 based on jdk-13+33)

This nightly version also shows the regression:

# "C:\dev\openj9_13_33\jdk-13+33\bin\java.exe" -version
openjdk version "13" 2019-09-17
OpenJDK Runtime Environment AdoptOpenJDK (build 13+33)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.17.0-m2, JRE 13 Windows 7 amd64-64-Bit Compressed References 20191009_89 (JIT enabled, AOT enabled)
OpenJ9   - 23b8b11b4
OMR      - 20db4fbc
JCL      - 2cb22d51d9 based on jdk-13+33)

But, this earlier Java 12 version does not show the regression:

# "C:\Program Files\AdoptOpenJDK\jdk-12.0.2.10-openj9\bin\java.exe" -version
openjdk version "12.0.2" 2019-07-16
OpenJDK Runtime Environment AdoptOpenJDK (build 12.0.2+10)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.15.1, JRE 12 Windows 7 amd64-64-Bit Compressed References 20190718_139 (JIT enabled, AOT enabled)
OpenJ9   - 0f66c6431
OMR      - ec782f26
JCL      - 06c2cc3322 based on jdk-12.0.2+10)

Summary of problem
------------------

I'm using an approximation of `Math.asin()` that I've written that is supposed to improve the performance of some trigonometry-heavy code, for projecting world maps onto globes. Using `Math.asin()` itself is a bad option here because I don't need meticulous precision (the globe is shown in a 300x300 pixel space), and to rotate at a decent speed it needs to reproject the map quickly. `Math.asin()` allows a single rotation, which includes quite a lot of non-trigonometry code, in 80ms or somewhat less. Using a fairly-precise approximation on HotSpot (any version since Java 8) makes asin() no longer a bottleneck, and it goes down to only needing 11 ms per rotation (which is quite acceptable). Using the same approximation on OpenJ9 with the non-regressed version above, 0.15.1, gives 14ms times rather reliably, which is also acceptable (there seems to be less jitter between frames, even though frames are slightly longer). Updating the OpenJ9 version to either of the regressed versions above, 0.16.0 and up, and running the same JAR gives 95-100ms times, which is not an acceptable speed for this application and introduces sharp jerking to the rotation. This means that with the regression, the approximation is now slower than the function it approximates, `Math.asin()`, and its performance is the bottleneck unless multiple other methods are slowing down at the same time. Microbenchmarks with BumbleBench confirm the empirical performance; they also show the approximation's speed has dropped below that of Math.asin(), and HotSpot runs the same code about 35x faster.

Diagnostic files
----------------

n/a

OutOfMemoryError: Java Heap Space
---------------------------------

n/a

Source code for approximation:
----------------------------

```java
////This was declared in the class NumberTools, so it is called as NumberTools.asin() in the bench.
    public static double asin(final double n)
    {
        final double ax = Math.sqrt(1.0 - n * n), ay = Math.abs(n);
        if(ax < ay)
        {
            final double a = ax / ay, s = a * a,
                    r = 1.57079637 - (((-0.0464964749 * s + 0.15931422) * s - 0.327622764) * s * a + a);
            return (n < 0.0) ? -r : r;
        }
        else {
            final double a = ay / ax, s = a * a,
                    r = (((-0.0464964749 * s + 0.15931422) * s - 0.327622764) * s * a + a);
            return (n < 0.0) ? -r : r;
        }
    }
```

Source code for BumbleBench benchmark:

```java
public final class ASinApproxBench extends MiniBench {
	protected int maxIterationsPerLoop(){ return 10000007; }

	protected long doBatch(long numLoops, int numIterationsPerLoop) throws InterruptedException {
		startTimer();
		double argument = NumberTools.asin(0.1);
		pauseTimer();
		double shrink = 1.99 / numIterationsPerLoop;
		for (long i = 0; i < numLoops; i++) {
			for (int j = 0; j < numIterationsPerLoop; j++) {
				startTimer();
				argument += NumberTools.asin(j * shrink - 1);
				pauseTimer();
			}
		}
		return numLoops * numIterationsPerLoop;
	}
}
```

Source code for the rotation demo is going to be harder to assemble; [you can see the code here](https://github.com/SquidPony/SquidLib/blob/master/squidlib-util/src/main/java/squidpony/squidgrid/mapping/WorldMapGenerator.java#L4402-L4499) and you can run [the pre-built JAR I mentioned (direct link)](https://github.com/tommyettinger/SquidLib-Demos/blob/master/Sampler/Sampler.jar?raw=true) as well. If you run the JAR, select RotatingWorldMapDemo in the dropdown and click Launch, then press the 'p' key to 'p'lay the animation when you're ready to observe timing. The time taken to generate a rotation is shown in the window titlebar; before you press 'p' it is a large number as it prepares the map content, and this number is larger on OpenJ9 0.16.0+ than other compilers (probably because it also uses `asin()`).

I'm hopeful that this will be something that can be tracked down based on when the regression happened, and that it will be feasible to reverse or mitigate. I want to use OpenJ9 for more things, especially because this type of performance decrease relative to HotSpot is rare. When it comes to certain pseudo-random number generators, OpenJ9 is blisteringly fast, and far outstrips Hotspot.

Let me know if there's anything I can do to help with diagnosing this.

**Comment by DanHeidinga** [comment]
fyi @andrewcraik

**Comment by tommyettinger** [comment]
Are there any general tips for how this could be tracked down (on my end or by an OpenJ9 dev)? I suspect this kind of performance drop doesn't happen often, because using released versions in the past hasn't resulted in any unexpected regressions. I say unexpected because going from Java 8 to Java 9 on Hotspot typically slows down everything a little, and I think the same was true for OpenJ9 to a lesser extent. I've got a little experience building a JDK from source, but not OpenJ9 yet; I think with a little help I could build OpenJ9 with whatever custom/debug information would be needed.

**Comment by smlambert** [comment]
@tommyettinger - I would like to suggest adding your benchmark to the set of [bumblebench tests](https://github.com/AdoptOpenJDK/openjdk-tests/tree/master/perf/bumbleBench) being added at AdoptOpenJDK (see https://github.com/AdoptOpenJDK/openjdk-tests/pull/1393 and https://github.com/AdoptOpenJDK/openjdk-tests/pull/1380 for similar).  We will begin to run those test targets regularly and hopefully will be able to catch (larger) regressions at the moment they are introduced.

**Comment by andrewcraik** [comment]
I have setup a bumblebench based on the supplied code and I have measured the performance with the Java 8 0.15.1 and 0.17 builds since this covers something close to the latest-and-greatest and the known good level reported for jdk12 while holding the Java language level and other variables constant across the releases. I can see a drop in performance of around 20% to 25% between the two releases when running the modified BumbleBench I created from the code in the issue.

A verbose log shows that roughly the same methods compiled at roughly the same hotness in both the 0.15.1 and 0.17 runs (note I ran -Xshareclasses:none to make sure that changes in AOT behavior were not part of the measurement). There is a single scorching method - the doBatch which would contain the math loop in question. I have generated logs for the compile in the two versions (0.15.1 and 0.17).

Checking the logs I can confirm that the inlining is the same across both 0.15.1 and 0.17 - we inline the same methods. Even the CFGs look visually the same and the optimization transformations also look nearly identical. I do see some differences in block ordering, but next step will be to grab some perf profiles of this thing to see if we can see where the time has gone because the optimizer is not doing something radically different between the two scorching compilations.

The second approach to this would be to narrow down the closest pair of nightly builds of OpenJ9 Java 8 the exhibit the performance drop - hopefully only one change is responsible for this and so we can find when that change was delivered to understand its impact on the system.

@tommyettinger thanks for the report and the test case. Fixing a performance issue like this is not something that is easy to write a standard guide for. My first step was to try and eliminate the different JDK lavels (eg using Java 8) to make sure the Java language level wasn't part of the problem. I will try and do a profile of the two cases to try and confirm where the time has gone. If you wanted to continue the investigation you could use the OpenJ9 nightly builds to see if you can find a pair where the drop occurred.

We appreciate you taking the time to submit the issue and develop all the tests and everything - it makes investigating the issue that bit easier. I am delighted you are enjoying OpenJ9 and thank you for giving us the chance to look into this issue.

FYI I removed the windows tag because I reproduced this on windows. I am not yet sure if this is platform specific so I have left the arch tag for now.

**Comment by andrewcraik** [comment]
Ok I've found it - in 0.15.1 we were generating some sqrtsd instruction for StrictMath.sqrt while we are not doing that for 0.17 - clearly a big problem. We'll need to see what happened in the codegen. FYI @fjeremic @simonameng @dchopra001 @SajinaKandy @BeverlyXu 

I see https://github.com/eclipse/openj9/pull/6217 which looks related...

**Comment by andrewcraik** [comment]
My reproduction was with
```
package net.adoptopenjdk.bumblebench.math;

import java.lang.Math;
import java.util.Random;

import net.adoptopenjdk.bumblebench.core.MiniBench;

public final class ASinApproxBench extends MiniBench {
        protected int maxIterationsPerLoop(){ return 100007; }

        public static double asin(final double n)
    {
        final double ax = Math.sqrt(1.0 - n * n), ay = Math.abs(n);
        if(ax < ay)
        {
            final double a = ax / ay, s = a * a,
                    r = 1.57079637 - (((-0.0464964749 * s + 0.15931422) * s - 0.327622764) * s * a + a);
            return (n < 0.0) ? -r : r;
        }
        else {
            final double a = ay / ax, s = a * a,
                    r = (((-0.0464964749 * s + 0.15931422) * s - 0.327622764) * s * a + a);
            return (n < 0.0) ? -r : r;
        }
    }

        protected long doBatch(long numLoops, int numIterationsPerLoop) throws InterruptedException {
                startTimer();
                double argument = asin(0.1);
                pauseTimer();
                double shrink = 1.99 / numIterationsPerLoop;
                for (long i = 0; i < numLoops; i++) {
                        for (int j = 0; j < numIterationsPerLoop; j++) {
                                startTimer();
                                argument += asin(j * shrink - 1);
                                pauseTimer();
                        }
                }
                return numLoops * numIterationsPerLoop;
        }
}
```

which I just dropped into the math directory of a bumblebench clone and re-ran ant dist

**Comment by andrewcraik** [comment]
Note I dropped the maxIterationsPerLoop to boost throughput.

**Comment by tommyettinger** [comment]
Woah, nice work @andrewcraik ! I wouldn't have expected that the difference could be so significant purely from one hardware-vs-software sqrt call, since it's called only once before a branching section that seems like it should take more time. Looking at my globe projection code, I also call `Math.sqrt()` an extra time per projected pixel, which means software sqrt would slow down both the asin() approximation and other code near where it is called. I wonder if the cause is simple or complex...

**Comment by fjeremic** [comment]
@simonameng and I are looking at the issue actively. As Andrew suspected the likely cause is #6217 which we'll confirm today and investigate what went wrong.

**Comment by andrewcraik** [comment]
Added the power tag based on the diff of the commit in #6217 it looks like power has the same problem.

**Comment by simonameng** [comment]
@andrewcraik @fjeremic  We think the problem is that in #6217, I tried to transform `StrictMath.sqrt` and `Math.sqrt` to `dsqrt` when `getSupportsHardwareSQRT` returns `true`. In that case, `inlineMathSQRT` won't be needed. So I deleted the `inlineMathSQRT` function on Z and X86 (Power doesn't have that function). However, I missed the fact that on X86 and Power, `getSupportsHardwareSQRT` will return `false` all the time, which means that X86 have to generate a call out to the VM which implements the method now, and it's causing the slow performance. 

In X86, I found the comment in [1] that the entry of `TR::dsqrt` should be `TR::TreeEvaluator::fpSqrtEvaluator`. But it seems that for now we're missing a table entry for it, and the function is not used anywhere. Because we always have a square root instruction available on X86, we don't need the processor to checks. We should implement a `dsqrt` evaluator and reduce the call the `dsqrt`. 

I am going to put the `inlineMathSQRT` function on X temporarily. And then we will work on a proper fix.

[1] https://github.com/eclipse/omr/blob/bfe4c5030f51573d21594b771410b0c21bf913c4/compiler/x/amd64/codegen/OMRTreeEvaluatorTable.hpp#L718

**Comment by andrewcraik** [comment]
I think I was mistaken about Power so I'm removing the tag.

**Comment by tommyettinger** [comment]
I can confirm with the October 30 nightly on Windows: not only is the regression gone, my asin benchmark is 10% faster with the latest nightly than with the OpenJ9 0.15.1 JDK before the regression. On my Windows laptop, the BumbleBench score with the latest nightly is 28197388.000000 (0.8% uncertainty), and with the older OpenJ9 0.15.1, it got a score of 25483476.000000 (0.9% uncertainty). With sufficient warmup, the globe projection demo I was running originally also gets comparable framerate to HotSpot, within 1ms per frame difference, which is unnoticeable at 60 frames per second.  Thank you; I'm very impressed!
